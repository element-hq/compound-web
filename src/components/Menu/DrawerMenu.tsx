/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, {
  type ComponentPropsWithoutRef,
  type ReactNode,
  forwardRef,
} from "react";
import styles from "./DrawerMenu.module.css";
import { getPlatform } from "../../utils/platform";
import classNames from "classnames";

interface Props extends ComponentPropsWithoutRef<"div"> {
  /**
   * The menu title.
   */
  title: string;
  /**
   * The menu contents.
   */
  children: ReactNode;
}

/**
 * A menu in a drawer, as commonly seen on mobile.
 */
// This an internal component not intended for export! Consumers should use it
// via the Menu or ContextMenu components.
export const DrawerMenu = forwardRef<HTMLDivElement, Props>(
  ({ title, children, className, ...props }, ref) => (
    <div
      ref={ref}
      className={classNames(className, styles.drawer)}
      aria-label={title}
      data-platform={getPlatform()}
      {...props}
      role="menu"
    >
      <div className={styles.body}>{children}</div>
    </div>
  ),
);

DrawerMenu.displayName = "DrawerMenu";
