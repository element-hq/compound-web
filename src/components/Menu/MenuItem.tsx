/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import classnames from "classnames";
import React, {
  ComponentPropsWithoutRef,
  ComponentType,
  ElementType,
  isValidElement,
  ReactElement,
  SVGAttributes,
  useCallback,
  useContext,
  MouseEventHandler,
} from "react";
import styles from "./MenuItem.module.css";
import { Text } from "../Typography/Text";
import ChevronRightIcon from "@vector-im/compound-design-tokens/assets/web/icons/chevron-right";
import { MenuContext } from "./MenuContext";
import { Slot } from "@radix-ui/react-slot";

type MenuItemElement = "button" | "a" | "div";

type Props<C extends MenuItemElement> = {
  /**
   * The element type of this menu item.
   * @default button
   */
  as?: C;
  /**
   * The CSS class name.
   */
  className?: string;
  /**
   * The icon to show on this menu item.
   * When `Icon` is a ReactElement, it should spread the props
   */
  Icon?: ComponentType<SVGAttributes<SVGElement>> | ReactElement;
  /**
   * The label to show on this menu item.
   */
  // This prop is required because it's rare to not want a label
  label: string | null;
  /**
   * Additional properties to pass to the Text label component.
   */
  labelProps?: ComponentPropsWithoutRef<typeof Text>;
  /**
   * Event callback for when the item is selected via mouse, touch, or keyboard.
   * Calling event.preventDefault in this handler will prevent the menu from
   * being dismissed.
   */
  // This prop is required because it's rare to not want a selection handler
  onSelect: ((e: Event) => void) | null;
  /**
   * Event callback for when the item is clicked.
   * @param e
   */
  onClick?: MouseEventHandler<HTMLElementTagNameMap[C]>;
  /**
   * The color variant of the menu item.
   * @default primary
   */
  kind?: "primary" | "critical";
  disabled?: boolean;
  /**
   * Whether to hide the chevron navigation hint.
   */
  hideChevron?: boolean;
} & Omit<ComponentPropsWithoutRef<C>, "onSelect" | "onClick">;

/**
 * An item within a menu, acting either as a navigation button, or simply a
 * container for other interactive elements.
 * Must be used within a compound Menu or other `menu` or `menubar` aria role subtree.
 */
export const MenuItem = <C extends MenuItemElement = "button">({
  as,
  className,
  Icon,
  label,
  labelProps,
  onSelect,
  kind = "primary",
  children,
  onClick: onClickProp,
  disabled,
  hideChevron,
  ...props
}: Props<C>): React.ReactElement => {
  const Component = as ?? ("button" as ElementType);
  const context = useContext(MenuContext);

  const onClick = useCallback(
    (e: Parameters<Exclude<typeof onClickProp, undefined>>[0]) => {
      (onClickProp as ((e_: typeof e) => void) | undefined)?.(e);
      // If there is no wrapper component to automatically handle onSelect, we
      // need to handle it manually, dismissing the menu as the default action
      if (onSelect !== null && context?.MenuItemWrapper == null) {
        const selectEvent = new CustomEvent("menu.itemSelect", {
          bubbles: true,
          cancelable: true,
        });
        onSelect(selectEvent);
        if (!selectEvent.defaultPrevented) context?.onOpenChange(false);
      }
    },
    [context, onSelect],
  );

  const iconIsReactElement = isValidElement(Icon);
  const componentIcon = Icon as ReactElement;
  const SvgIcon = Icon as ComponentType<SVGAttributes<SVGElement>>;

  const content = (
    <Component
      role="menuitem"
      {...props}
      className={classnames(className, styles.item, {
        [styles.interactive]: onSelect !== null,
        [styles["no-label"]]: label === null,
        [styles["no-icon"]]: !Icon,
        [styles["disabled"]]: disabled,
      })}
      data-kind={kind}
      onClick={onClick}
      disabled={Component === "button" ? disabled : undefined}
      aria-disabled={Component === "button" ? undefined : disabled}
    >
      {Icon &&
        (iconIsReactElement ? (
          <Slot className={styles.icon}>{componentIcon}</Slot>
        ) : (
          <SvgIcon
            width={24}
            height={24}
            className={styles.icon}
            aria-hidden={true}
          />
        ))}

      {label !== null && (
        <Text
          className={styles.label}
          size="md"
          weight="medium"
          as="span"
          {...labelProps}
        >
          {label}
        </Text>
      )}
      {/* We use CSS to swap between this navigation hint and the provided
      children on hover - see the styles module. */}
      {!hideChevron && (Component === "button" || Component === "a") && (
        <ChevronRightIcon
          width={8}
          height={24}
          className={styles["nav-hint"]}
          aria-hidden={true}
          /* The SVG is a small icon in a large canvas. It probably ought to be
             cropped, but we can adjust the viewBox here to chop off the horizontal
             space to get it closer to the right hand edge.
           */
          viewBox="8 0 8 24"
        />
      )}
      {children}
    </Component>
  );

  return context?.MenuItemWrapper == null || onSelect === null ? (
    content
  ) : (
    <context.MenuItemWrapper onSelect={onSelect}>
      {content}
    </context.MenuItemWrapper>
  );
};
