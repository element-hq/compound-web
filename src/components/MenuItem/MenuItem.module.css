/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.item {
  display: grid;
  grid-template: "icon label ." auto "empty1 label empty2" auto / auto auto minmax(
      /* Reserve space for the chevron so that the layout doesn't shift on
      hover */
        var(--cpd-space-4x),
      1fr
    );
  align-items: center;
  justify-items: end;
  padding-block: var(--cpd-space-2x);
  padding-inline: var(--cpd-space-4x);
  box-sizing: border-box;
  inline-size: 100%;
  min-inline-size: 200px;
  color: var(--cpd-color-text-secondary);
  background: var(--cpd-color-bg-action-secondary-rest);
}

.item.interactive {
  cursor: pointer;
}

.item.no-label {
  grid-template: "icon ." auto / auto 1fr;
}

/* Items with interactive controls like toggles are larger by default */
label.item {
  min-inline-size: 246px;
}

.icon {
  grid-area: icon;
  margin-inline-end: var(--cpd-space-3x);
}

.item.no-label .icon {
  margin-inline-end: var(--cpd-space-4x);
}

.label {
  grid-area: label;
  margin-inline-end: var(--cpd-space-4x);
  text-align: start;
  word-break: break-word;
}

.nav-hint {
  /* Hidden until the item is hovered over */
  display: none;
  flex-shrink: 0;
  margin-inline-end: calc(-1 * var(--cpd-space-2x));
}

button.item {
  appearance: none;
  border: none;
}

.item[data-kind="primary"] > .label {
  color: var(--cpd-color-text-primary);
}

.item[data-kind="primary"] > .icon {
  color: var(--cpd-color-icon-primary);
}

.item[data-kind="primary"] > .nav-hint {
  color: var(--cpd-color-icon-tertiary);
}

.item[data-kind="critical"] > .label {
  color: var(--cpd-color-text-critical-primary);
}

.item[data-kind="critical"] > .icon,
.item[data-kind="critical"] > .nav-hint {
  color: var(--cpd-color-icon-critical-primary);
}

@media (hover) {
  .item.interactive[data-kind="primary"]:hover {
    background: var(--cpd-color-bg-action-secondary-hovered);
  }

  .item.interactive[data-kind="critical"]:hover {
    background: var(--cpd-color-bg-critical-subtle);
  }

  /* Replace the children with the navigation hint on hover */
  .item.interactive:hover > .nav-hint {
    display: initial;
  }

  .item.interactive:hover > .nav-hint ~ * {
    display: none;
  }
}

.item.interactive[data-kind="primary"]:active {
  background: var(--cpd-color-bg-action-secondary-pressed);
}

.item.interactive[data-kind="critical"]:active {
  background: var(--cpd-color-bg-critical-subtle);
}

.item[data-kind].disabled {
  pointer-events: none;
}

.item[data-kind].disabled > .label,
.item[data-kind].disabled > .icon,
.item[data-kind].disabled > .nav-hint {
  color: var(--cpd-color-text-disabled);
}
