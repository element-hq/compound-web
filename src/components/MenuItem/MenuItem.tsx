/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, {
  ComponentPropsWithoutRef,
  ComponentType,
  ElementType,
  SVGAttributes,
} from "react";
import styles from "./MenuItem.module.css";
import { Text } from "../Typography/Text";
import ChevronRightIcon from "@vector-im/compound-design-tokens/icons/chevron-right.svg";

type MenuItemElement = "button" | "label" | "a" | "div";

type Props<C extends MenuItemElement> = {
  /**
   * The element type of this menu item.
   * @default button
   */
  as?: C;
  /**
   * The CSS class name.
   */
  className?: string;
  /**
   * The icon to show on this menu item.
   */
  Icon: ComponentType<SVGAttributes<SVGElement>>;
  /**
   * The label to show on this menu item.
   */
  // This prop is required because it's rare to not want a label
  label: string | undefined;
  /**
   * The color variant of the menu item.
   * @default primary
   */
  kind?: "primary" | "critical";

  disabled?: boolean;
} & ComponentPropsWithoutRef<C>;

/**
 * An item within a menu, acting either as a navigation button, or simply a
 * container for other interactive elements.
 */
export const MenuItem = <C extends MenuItemElement = "button">({
  as,
  className,
  Icon,
  label,
  kind = "primary",
  children,
  disabled,
  ...props
}: Props<C>): JSX.Element => {
  const Component = as ?? ("button" as ElementType);

  return (
    <Component
      role="menuitem"
      {...props}
      className={classnames(className, styles.item, {
        [styles.interactive]: as !== "div",
        [styles["no-label"]]: label === undefined,
        [styles["disabled"]]: disabled,
      })}
      data-kind={kind}
      disabled={disabled}
    >
      <Icon width={24} height={24} className={styles.icon} aria-hidden={true} />
      {label !== undefined && (
        <Text className={styles.label} size="md" weight="medium" as="span">
          {label}
        </Text>
      )}
      {/* We use CSS to swap between this navigation hint and the provided
      children on hover - see the styles module. */}
      {(Component === "button" || Component === "a") && (
        <ChevronRightIcon
          width={24}
          height={24}
          className={styles["nav-hint"]}
          aria-hidden={true}
        />
      )}
      {children}
    </Component>
  );
};
