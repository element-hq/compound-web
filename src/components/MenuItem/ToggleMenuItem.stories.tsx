/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";
import ExtensionsIcon from "@vector-im/compound-design-tokens/icons/extensions.svg";
import ChatIcon from "@vector-im/compound-design-tokens/icons/chat.svg";

import { ToggleMenuItem as ToggleMenuItemComponent } from "./ToggleMenuItem";

export default {
  title: "ToggleMenuItem",
  component: ToggleMenuItemComponent,
  tags: ["autodocs"],
  argTypes: {},
  args: {},
} as Meta<typeof ToggleMenuItemComponent>;

const Template: StoryFn<typeof ToggleMenuItemComponent> = (args) => (
  <div style={{ width: 300 }}>
    <ToggleMenuItemComponent {...args} Icon={ChatIcon} label="First item" />
    <ToggleMenuItemComponent
      {...args}
      Icon={ExtensionsIcon}
      label="Second item with a name that's quite long"
    />
  </div>
);

export const Primary = Template.bind({});
Primary.args = {};
