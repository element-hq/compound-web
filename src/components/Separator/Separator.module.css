/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.separator {
  --cpd-separator-spacing: var(--cpd-space-2x);

  background-color: var(--cpd-color-gray-400);
}

.separator[data-orientation="horizontal"] {
  margin-block: var(--cpd-separator-spacing);
  block-size: 1px;
  inline-size: 100%;
}

.separator[data-orientation="vertical"] {
  margin-inline: var(--cpd-separator-spacing);
  block-size: 100%;
  inline-size: 1px;
}
