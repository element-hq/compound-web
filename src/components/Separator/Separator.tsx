/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import * as RadixSeparator from "@radix-ui/react-separator";
import React, {
  ComponentProps,
  PropsWithoutRef,
  Ref,
  forwardRef,
  useEffect,
} from "react";
import styles from "./Separator.module.css";

type SeparatorProps = {
  /**
   * The CSS class name.
   */
  className?: string;
  /**
   * The spacing on either side of the separator
   */
  spacing?: string;
} & ComponentProps<typeof RadixSeparator.Root>;

const SPACING_CUSTOM_PROP = "--cpd-separator-spacing";

/**
 * A separator component
 * Note: Only supports `React.RefObject`
 */
export const Separator = forwardRef(
  (
    {
      className,
      spacing = "var(--cpd-space-2x)",
      ...props
    }: PropsWithoutRef<SeparatorProps>,
    ref?: Ref<HTMLDivElement>,
  ) => {
    const classes = classnames(styles.separator, className);

    useEffect(() => {
      // Casting to `RefObject` as everything here should use this type.
      // Functions refs are not something we wish to support and string refs
      // are already unsupported with `forwardRef`.
      const style = (ref as React.RefObject<HTMLDivElement>)?.current?.style;
      if (spacing) {
        style?.setProperty(SPACING_CUSTOM_PROP, spacing);
      } else {
        style?.removeProperty(SPACING_CUSTOM_PROP);
      }
    }, [spacing]);

    return <RadixSeparator.Root {...props} className={classes} ref={ref} />;
  },
);

Separator.displayName = "Separator";
