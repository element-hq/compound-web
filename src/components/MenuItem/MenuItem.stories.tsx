/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";
import ExtensionsIcon from "@vector-im/compound-design-tokens/icons/extensions.svg";
import ChatIcon from "@vector-im/compound-design-tokens/icons/chat.svg";
import SettingsLabel from "@vector-im/compound-design-tokens/icons/settings.svg";

import { MenuItem as MenuItemComponent } from "./MenuItem";
import { Text } from "../Typography/Text";

export default {
  title: "MenuItem",
  component: MenuItemComponent,
  tags: ["autodocs"],
  argTypes: {
    onClick: { action: "clicked!" },
  },
  args: {
    label: "Menu item",
    children: (
      <Text as="span" size="sm">
        99
      </Text>
    ),
    Icon: ChatIcon,
  },
  decorators: [
    (Story: StoryFn) => (
      <div style={{ width: 300 }}>
        <Story />
      </div>
    ),
  ],
} as Meta<typeof MenuItemComponent>;

export const Example = {
  render: () => (
    <div style={{ width: 300 }}>
      <MenuItemComponent Icon={ChatIcon} label="First item">
        <Text as="span" size="sm">
          99
        </Text>
      </MenuItemComponent>
      <MenuItemComponent
        Icon={ExtensionsIcon}
        label="Second item with a name that's quite long"
      />
      <MenuItemComponent Icon={SettingsLabel} label={undefined}>
        <Text as="span" size="sm">
          Third item without a label
        </Text>
      </MenuItemComponent>
    </div>
  ),
};

export const Primary = {
  args: {
    kind: "primary",
  },
};

export const Critical = {
  args: {
    kind: "critical",
  },
};

export const PrimaryDisabled = {
  args: {
    ...Primary.args,
    disabled: true,
  },
};

export const CriticalDisabled = {
  args: {
    ...Critical.args,
    disabled: true,
  },
};

export const WithoutLabel = {
  args: {
    ...Primary.args,
    label: undefined,
  },
};

export const WithALongLabel = {
  args: {
    ...Primary.args,
    label:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
  },
};

export const WithALongLabelAndChildren = {
  args: {
    ...Primary.args,
    label:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, empor incididunt ut labore et dolore magna aliqua.",
    children: (
      <Text as="span" size="sm">
        Longer children too
      </Text>
    ),
  },
};
