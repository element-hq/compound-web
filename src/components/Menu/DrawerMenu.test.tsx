/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import React from "react";
import UserProfileIcon from "@vector-im/compound-design-tokens/assets/web/icons/user-profile";
import LeaveIcon from "@vector-im/compound-design-tokens/assets/web/icons/leave";

import { DrawerMenu } from "./DrawerMenu";
import { MenuItem } from "./MenuItem";
import { Separator } from "../Separator/Separator";

describe("DrawerMenu", () => {
  it("renders", () => {
    const { asFragment } = render(
      <DrawerMenu title="Settings">
        <MenuItem Icon={UserProfileIcon} label="Profile" onSelect={() => {}} />
        <Separator />
        <MenuItem
          kind="critical"
          Icon={LeaveIcon}
          label="Sign out"
          onSelect={() => {}}
        />
      </DrawerMenu>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
