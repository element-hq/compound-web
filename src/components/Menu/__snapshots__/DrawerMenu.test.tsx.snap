// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DrawerMenu > renders 1`] = `
<DocumentFragment>
  <div
    aria-label="Settings"
    class="_drawer_31bf0e"
    data-platform="other"
    role="menu"
  >
    <div
      class="_body_31bf0e"
    >
      <button
        class="_item_831119 _interactive_831119"
        data-kind="primary"
        role="menuitem"
      >
        <svg
          aria-hidden="true"
          class="_icon_831119"
          fill="currentColor"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.175 13.825Q10.35 15 12 15t2.825-1.175T16 11t-1.175-2.825T12 7 9.175 8.175 8 11t1.175 2.825m4.237-1.412A1.93 1.93 0 0 1 12 13q-.825 0-1.412-.588A1.93 1.93 0 0 1 10 11q0-.825.588-1.412A1.93 1.93 0 0 1 12 9q.825 0 1.412.588Q14 10.175 14 11t-.588 1.412"
          />
          <path
            d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10m-2 0a8 8 0 1 0-16 0 8 8 0 0 0 16 0"
          />
          <path
            d="M16.23 18.792a13 13 0 0 0-1.455-.455 11.6 11.6 0 0 0-5.55 0q-.73.18-1.455.455a8 8 0 0 1-1.729-1.454q1.336-.618 2.709-.95A13.8 13.8 0 0 1 12 16q1.65 0 3.25.387 1.373.333 2.709.95a8 8 0 0 1-1.73 1.455"
          />
        </svg>
        <span
          class="_typography_489030 _font-body-md-medium_489030 _label_831119"
        >
          Profile
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_831119"
          fill="currentColor"
          height="24"
          viewBox="8 0 8 24"
          width="8"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.7 17.3a.95.95 0 0 1-.275-.7q0-.425.275-.7l3.9-3.9-3.9-3.9a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l4.6 4.6q.15.15.213.325.062.175.062.375t-.062.375a.9.9 0 0 1-.213.325l-4.6 4.6a.95.95 0 0 1-.7.275.95.95 0 0 1-.7-.275"
          />
        </svg>
      </button>
      <div
        class="_separator_162edc"
        data-kind="primary"
        data-orientation="horizontal"
        role="separator"
      />
      <button
        class="_item_831119 _interactive_831119"
        data-kind="critical"
        role="menuitem"
      >
        <svg
          aria-hidden="true"
          class="_icon_831119"
          fill="currentColor"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 13q.424 0 .713-.287A.97.97 0 0 0 15 12a.97.97 0 0 0-.287-.713A.97.97 0 0 0 14 11a.97.97 0 0 0-.713.287A.97.97 0 0 0 13 12q0 .424.287.713.288.287.713.287"
          />
          <path
            d="M10.385 21.788A1 1 0 0 1 10 21V3a1.003 1.003 0 0 1 1.242-.97l8 2A1 1 0 0 1 20 5v14a1 1 0 0 1-.758.97l-8 2a1 1 0 0 1-.857-.182M18 5.781l-6-1.5v15.438l6-1.5zM9 6H7v12h2v2H7a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h2z"
          />
        </svg>
        <span
          class="_typography_489030 _font-body-md-medium_489030 _label_831119"
        >
          Sign out
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_831119"
          fill="currentColor"
          height="24"
          viewBox="8 0 8 24"
          width="8"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.7 17.3a.95.95 0 0 1-.275-.7q0-.425.275-.7l3.9-3.9-3.9-3.9a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l4.6 4.6q.15.15.213.325.062.175.062.375t-.062.375a.9.9 0 0 1-.213.325l-4.6 4.6a.95.95 0 0 1-.7.275.95.95 0 0 1-.7-.275"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
